src/javascript/note.js

```diff
@@ -1,94 +1,215 @@
-export default function Note(vm, x, width, color) {
-  this.x = x;
-  this.width = width;
-  this.color = color;
-  const { ctx, canvas } = vm;
+export default class Note {
+  constructor(vm, game, keyObj, key, x, width, color) {
+    this.x = x;
+    this.width = width;
+    this.color = color;
+    this.vm = vm;
+    this.game = game;
+    this.keyObj = keyObj;
+    this.key = key;
+    this.ctx = vm.ctx;
+    this.canvas = vm.canvas;
+    this.percentage = 0;
+    this.isHoldNote = keyObj.h && keyObj.h[key];
+    this.isUserHolding = false;
+    this.isHoldingDone = false;
 
-  this.y = 0;
+    this.y = 0;
+    this.singleNoteHeight = 10;
 
-  // modulate speed, ref https://www.viget.com/articles/time-based-animation/
-  this.now = Date.now();
-  this.delta = 0;
-  this.then = 0;
-  this.timeStarted = this.now;
+    // modulate speed, ref https://www.viget.com/articles/time-based-animation/
+    this.now = Date.now();
+    this.delta = 0;
+    this.then = 0;
+    this.timeStarted = this.now;
+  }
 
-  this.setDelta = function () {
+  setDelta() {
     if (this.then === 0) this.then = this.now;
     this.now = Date.now();
     this.delta = (parseFloat(this.now) - parseFloat(this.then)) / 1000; // seconds since last frame
     this.then = this.now;
-  };
+  }
+
+  getDiffPercentage() {
+    if (this.gameHadBeenPaused) {
+      // game had been pasued, time unuseable, use less accurate dist calculation instead
+      const dist = this.vm.checkHitLineY - this.y;
+      const percentage = Math.abs(dist) / this.canvas.height; // the lower the better
+
+      return percentage;
+    } else {
+      let hitTimeSinceStartInSec =
+        (parseFloat(Date.now()) - parseFloat(this.timeStarted)) / 1000;
+      const diff = Math.abs(this.vm.noteSpeedInSec - hitTimeSinceStartInSec);
+      let percentage = diff / this.vm.noteSpeedInSec; // the lower the better
 
-  this.getDiffPercentage = function () {
-    // const dist = vm.checkHitLineY - this.y;
-    // const percentage = Math.abs(dist) / (canvas.height / 10); // the lower the better
-    let hitTimeSinceStartInSec =
-      (parseFloat(Date.now()) - parseFloat(this.timeStarted)) / 1000;
-    const diff = Math.abs(vm.noteSpeedInSec - hitTimeSinceStartInSec);
-    let percentage = diff / vm.noteSpeedInSec; // the lower the better
+      return percentage;
+    }
+  }
+
+  calculatePercent() {
+    this.percentage = 1 - ((1 - this.getDiffPercentage() - 0.2) / 4) * 5;
+    this.percentage = this.percentage < 1 ? this.percentage : 1;
```

Comment:
This scoring function doesn't have range ∈ [0, 1] does it lol

Reply:
It's a mess here...